### PR TITLE
Adding type hints to deep_lift.py

### PR DESCRIPTION
**Pull request by kolvia**
Adding type hints to:
```
captum/attr/_core/deep_lift.py
tests/attr/test_deeplift_basic.py
tests/attr/test_deeplift_classification.py
```

I've added a number of type hints so far but I'm running into issues with what's left. The biggest is that whenever I try to add type hints to `DeepLiftShap.attribute`, I get: `error: Signature of "attribute" incompatible with supertype "DeepLift"`. So, just wanted to get some feedback on what I have so far and see if there are any suggestions going forward.

### CODE REVIEW

File: captum/attr/_core/deep_lift.py [hunk 1]
```diff
@@ -55,16 +60,57 @@ def is_output_cloned(output_fn, input_grad_fn):
...
         r"""
         Args:
 
             model (nn.Module):  The reference to PyTorch model instance.
         """
         GradientAttribution.__init__(self, model)
         self.model = model
-        self.forward_handles = []
-        self.backward_handles = []
+        self.forward_handles: List[Any] = []
```
Comment (vivekmig) [inline captum/attr/_core/deep_lift.py]: This is more precisely a List of RemovableHandle (torch.utils.hooks).

File: captum/attr/_core/deep_lift.py [hunk 2]
```diff
@@ -310,10 +356,10 @@ def forward_fn():
             forward_fn.device_ids = forward_func.device_ids
         return forward_fn
 
-    def _is_non_linear(self, module):
+    def _is_non_linear(self, module: nn.Module) -> bool:
         return type(module) in SUPPORTED_NON_LINEAR.keys()
 
-    def _forward_pre_hook_ref(self, module, inputs):
+    def _forward_pre_hook_ref(self, module: Any, inputs: TensorOrTupleOfTensors) -> None:
```
Comment (vivekmig) [inline captum/attr/_core/deep_lift.py]: This module should also be Module, note that this may require another "# type: ignore", since we add an attribute to the Module, which is not typed in the Module.

File: tests/attr/test_deeplift_basic.py [hunk 3]
```diff
@@ -263,8 +275,13 @@ def test_lin_maxpool_lin_classification(self):
         assertArraysAlmostEqual(delta.detach().numpy(), expected_delta)
 
     def _deeplift_assert(
-        self, model, attr_method, inputs, baselines, custom_attr_func=None
-    ):
+        self,
+        model: Module,
+        attr_method: Union[DeepLift, DeepLiftShap],
+        inputs: Tuple[Tensor, ...],
+        baselines,  # ?
```
Comment (vivekmig) [inline tests/attr/test_deeplift_basic.py]: You can leave out type hinting this argument, it's a little messy since it's used by both DeepLift and DeepLiftShap, but they don't have consistent baseline arguments.

**Comment by vivekmig** [comment]
Also, it looks like there's some lint issues, so once you finish your changes, just run black and flake8 to address linter issues. Make sure these are upgraded to the latest versions, since black may make extraneous changes to other files otherwise.

**Comment by kolvia** [comment]
@vivekmig Thanks for the help! 

I've added some more type hinting (and fixed the linter issues) but I'm running into more errors:

`captum/attr/_core/deep_lift.py:499: error: Argument 1 to "register_backward_hook" of "Module" has incompatible type "Callable[[Module[Any], TensorOrTupleOfTensors, TensorOrTupleOfTensors, float], Any]"; expected "Callable[[Module[Any], Union[Tuple[Tensor, ...], Tensor], Union[Tuple[Tensor, ...], Tensor]], Optional[Tensor]]"
captum/attr/_core/deep_lift.py:841: error: Incompatible types in assignment (expression has type "Tuple[Any, ...]", variable has type "Tuple[int, int]")
captum/attr/_core/deep_lift.py:842: error: No overload variant of "mean" matches argument types "Tensor", "int", "bool"
captum/attr/_core/deep_lift.py:842: note: Possible overload variants:
captum/attr/_core/deep_lift.py:842: note:     def mean(self: Tensor, *, dtype: Optional[dtype] = ..., out: Optional[Tensor] = ...) -> Tensor
captum/attr/_core/deep_lift.py:842: note:     def mean(self: Tensor, dim: Union[int, Union[Size, List[int], Tuple[int, ...]]], keepdim: bool = ..., *, dtype: Optional[dtype] = ..., out: Optional[Tensor] = ...) -> Tensor
captum/attr/_core/deep_lift.py:842: note:     def mean(self: Tensor, dim: List[Optional[str]], keepdim: bool = ..., *, dtype: Optional[dtype] = ..., out: Optional[Tensor] = ...) -> Tensor
captum/attr/_core/deep_lift.py:992: error: Argument 1 to "list" has incompatible type "Union[Size, Tensor, Module[Any]]"; expected "Iterable[Any]"
captum/attr/_core/deep_lift.py:1010: error: Argument 1 to "list" has incompatible type "Union[Size, Tensor, Module[Any]]"; expected "Iterable[Any]"
captum/attr/_core/neuron/neuron_deep_lift.py:186: error: Incompatible return value type (got "Union[Tensor, Tuple[Tensor, Tensor]]", expected "Tensor")
captum/attr/_core/neuron/neuron_deep_lift.py:186: error: Incompatible return value type (got "Union[Tuple[Tensor, ...], Tuple[Tuple[Tensor, ...], Tensor]]", expected "Tuple[Tensor, ...]")
captum/attr/_core/neuron/neuron_deep_lift.py:360: error: Argument 2 to "attribute" of "DeepLiftShap" has incompatible type "Union[float, Tensor, Tuple[Union[Tensor, float], ...], Callable[..., Union[Tensor, Tuple[Tensor, ...]]], None]"; expected "Union[Tensor, Callable[..., Tensor]]"
captum/attr/_core/neuron/neuron_deep_lift.py:360: error: Argument 2 to "attribute" of "DeepLiftShap" has incompatible type "Union[float, Tensor, Tuple[Union[Tensor, float], ...], Callable[..., Union[Tensor, Tuple[Tensor, ...]]], None]"; expected "Union[Tuple[Tensor, ...], Callable[..., Tuple[Tensor, ...]]]"
Found 9 errors in 2 files (checked 44 source files)`

I'm kind of stuck on how to fix these. Additionally, there's still some functions I haven't finished adding types to (`_check_valid_module` and the maxpool functions) where I'm not really sure what types they take.

File: captum/attr/_core/deep_lift.py [hunk 4]
```diff
@@ -792,7 +950,14 @@ def maxpool3d(module, inputs, outputs, grad_input, grad_output, eps=1e-10):
...
-    module, pool_func, unpool_func, inputs, outputs, grad_input, grad_output, eps=1e-10,
+    module: Module,
+    pool_func: Callable,
+    unpool_func: Callable,
+    inputs,
+    outputs,
+    grad_input,
+    grad_output,
+    eps: float = 1e-10,
 ):
```
Comment (vivekmig) [inline captum/attr/_core/deep_lift.py]: This method has two instances of `list(module.input.shape)`, you can replace these with `list(cast(torch.Size, module.input.shape))`

File: captum/attr/_core/deep_lift.py [hunk 5]
```diff
@@ -55,16 +62,57 @@ def is_output_cloned(output_fn, input_grad_fn):
...
-        self.backward_handles = []
+        self.forward_handles: List[RemovableHandle] = []
+        self.backward_handles: List[RemovableHandle] = []
+
+    @typing.overload
+    def attribute(
+        self,
+        inputs: TensorOrTupleOfTensors,
+        baselines: Optional[
+            Union[int, Tuple[int, ...], Tensor, List[Tuple[int, ...]]]
```
Comment (vivekmig) [inline captum/attr/_core/deep_lift.py]: This baselines hint should match the ones below, `Union[Tensor, int, float, Tuple[Union[Tensor, int, float], ...]]`

**Comment by kolvia** [comment]
@vivekmig I think it should be good to go, now. Thanks so much for all your help!

File: captum/attr/_core/deep_lift.py [hunk 6]
```diff
@@ -427,14 +488,16 @@ def _register_hooks(self, module):
...
         for forward_handle in self.forward_handles:
             forward_handle.remove()
         for backward_handle in self.backward_handles:
             backward_handle.remove()
 
-    def _pre_hook_main_model(self):
-        def pre_hook(module, baseline_inputs_add_args):
+    def _pre_hook_main_model(self) -> RemovableHandle:
+        def pre_hook(
+            module: Module, baseline_inputs_add_args: Tuple[Tensor, ...]
```
Comment (vivekmig) [inline captum/attr/_core/deep_lift.py]: Since the tuple includes additional arguments which could be anything, the type of both baseline_inputs_add_args and the return type should be just Tuple.

File: tests/attr/test_deeplift_basic.py [hunk 7]
```diff
@@ -209,8 +213,12 @@ def gen_baselines_returns_array():
         assertTensorAlmostEqual(self, attributions[0], attributions_with_func[0])
         assertTensorAlmostEqual(self, attributions[1], attributions_with_func[1])
 
-    def test_relu_deepliftshap_with_custom_attr_func(self):
-        def custom_attr_func(multipliers, inputs, baselines):
+    def test_relu_deepliftshap_with_custom_attr_func(self) -> None:
+        def custom_attr_func(
+            multipliers: Tuple[Tensor, ...],
+            inputs: Tuple[Tensor, ...],
+            baselines: Tuple[Union[Tensor, int, float], ...],
```
Comment (vivekmig) [inline tests/attr/test_deeplift_basic.py]: This type hint for baselines should actually be Tuple[Tensor, ...], since we internally convert to a Tensor before this method is called.
